packages/gatsby/src/query/query-compiler.js

```diff
@@ -183,74 +173,37 @@ class Runner {
         if (definition.kind === Kind.FRAGMENT_DEFINITION) {
           const fragmentName = definition.name.value
           if (fragmentMap.has(fragmentName)) {
-            if (print(definition) === fragmentMap.get(fragmentName)) {
+            if (print(definition) === fragmentMap.get(fragmentName).text) {
               return false
             }
           } else {
-            fragmentMap.set(fragmentName, print(definition))
+            fragmentMap.set(fragmentName, {
+              def: definition,
+              text: print(definition),
+            })
           }
         }
         return true
       })
 
-      documents.push(doc)
       doc.definitions.forEach((def: any) => {
         const name: string = def.name.value
```

Comment:
Maybe we could merge this loop with a filter above (as this filter has side effects anyway)?

Reply:
Good idea!